Gdpr and app security (#1154)
oktadeveloper/okta-blog: _source/_posts/2022-04-28-gdpr-and-app-security.md

---
type: awareness
layout: blog_post

> title: "GDPR and application security"

Reviewer: ```suggestion
title: "The Benefits of GDPR for Application Security"
```

> github: https://github.com/oktadev/okta-dotnet6-webapi-example

Reviewer: Move `type` to bottom and remove GitHub.

> title: "GDPR and application security"

Reviewer: We should probably use a more descriptive title with keywords. Maybe use Keyword Planner and https://blog-title-analyzer.oktadev.dev to figure out a better title. Also, it should be title case. You can use this site for help. http://www.conissaunce.com/pages/infoq-title-case.html

Author: Hmm I thought these probably were already proofed by Havi. May be not, I'll update

Reviewer 2: "The Benefits of GDPR for Application Security" scores very well!

Reviewer: Nice. I'll update it.

author: olabayo-balogun
by: contractor

> communities: []

Reviewer: ```suggestion
communities: [security]
```

tweets:
  - "GDPR rules can be long and esoteric, but it's important to know how to stay in compliance with the policy. Not only will you avoid fines for violations, but users will be more willing to trust your application. In other words: compliance is worth the work."
  - "As the internet has grown, so too has the need to protect users from privacy invasions, fraud, and other types of abuse. If your organization or application is subject to GDPR rules, you'll need to know how to get and stay in compliance."
image: blog/gdpr-and-app-security/gdpr-cover.jpg
---

As the internet has grown, so has the need to protect users from privacy invasions, fraud, or other types of abuse by attackers. The European Union's solution is the General Data Protection Regulation, or [GDPR](https://gdpr.eu/). First put into effect in 2018, the policy applies not only to the organizations in the EU but to any organization that uses or collects data from people in the EU.

If your organization is subject to GDPR, you'll need to know how to stay in compliance. This article will break down the details of this policy so that you understand its potential effects on your application.

> 

Reviewer: ```suggestion

{% include toc.md %}
```

## What to know about GDPR

> GDPR, which officially took effect on May 25, 2018, primarily encourages organizations to properly handle user data and privacy. Those not in compliance can face fines, in some cases [large ones](https://gdpr.eu/the-gdpr-meets-its-first-challenge-facebook/).

Reviewer: "large ones like Facebook" might be better to give the reader an indication of what they're clicking on.

### GDPR requirements

GDPR compliance imposes multiple requirements on your organization. Here is what you must include in your application:

> - Enhanced application security.

Reviewer: Can these be summarized? For example:

```
- Enhanced application security (e.g. HTTPS, Foo, Bar, etc).
```

Author: GDPR doesn't explicity define this, the closest would be a vague definition that requires end to end encryption and 2FA

## How GDPR affects application security

If you're responsible for managing your application's security, GDPR can feel like a threat. There is a lot of work involved, and mistakes can lead to costly fines. If you use it properly, though, following GDPR can help your application. Here are some of the benefits it can offer:

### Bigger security budget

Prior to GDPR, many organizations viewed cybersecurity as an expense rather than an investment. This attitude was reflected in understaffing and decreased purchasing power for critical cybersecurity features. The result was easily exploited digital infrastructure.

Because GDPR demands that organizations increase data security and backs up those demands with fines, managers of organizations would rather devote more of their budget to cybersecurity than risk losing profit to a security breach.

### More security integrations

> Applications that help with [automation testing](https://www.globalapptesting.com/blog/what-is-automation-testing), [automated breach and attack simulation](https://www.xmcyber.com/what-is-breach-and-attack-simulation/), [virtual private networks (VPNs)](https://www.kaspersky.com/resource-center/definitions/what-is-a-vpn), and [virtual private clouds (VPCs)](https://www.ibm.com/cloud/learn/vpc) are more likely to be added to an organization's security strategy. Such additions help shore up the defenses of digital infrastructure; VPCs and VPNs also offer a sandbox environment for employees to experiment with security features.

Reviewer: This is kinda confusing because this post is about application security, which seems to target me as a developer. However, "Applications" in this sentence seems to be more like products you'd buy to enhance your infrastructure's security.

### Data mapping

[Data mapping](https://en.wikipedia.org/wiki/Data_mapping), or connecting information from multiple data sets, requires a top-down analysis of all of an organization's databases and what information they hold. This can be a big task, but it ensures transparency and accountability. Data mapping is required for the next steps.

Data mapping can be manual, semi-automated, or automated. You should determine which approach works best for you, depending on the size of your company, your budget, the quality of your data management team, and the amount of data you work with. Deciding how to [implement data mapping](https://blog.hubspot.com/marketing/data-mapping) will require some understanding of your databases, as well as knowing which duplicated columns can be optimized to reduce redundancy.

### Data purge functionality

Previously, even after users requested that their data be deleted, organizations could remove it from public access and keep it somewhere hidden; however, the GDPR mandates that "officially" deleted data must be deleted everywhere. Organizations must ensure that they monitor data to delete it properly from all resources as needed.

> Data purge functionality [implementation](https://success.outsystems.com/Documentation/Best_Practices/Architecture/Data_Purging) will likely require some modification to your database schema so that you can identify data to mark for deletion, per user requests, and data to purge due to staleness. A `LastUpdatedDate` and `IsDeleted` column in your data table can be useful in identifying either type of data.

Reviewer: This sentence reads funny. It seems like it should be something like "A data purge implementation..."

Reviewer: > A `LastUpdatedDate` and `IsDeleted` column

This seems like it should be plural. For example, "Using a `last_updated_date` or `is_deleted` columns..."

I changed the column names because who uses CamelCase for database column names?!

Author: Well depends on DB 😜

### Refactoring DB schemas

> Data purge functionality may involve refactoring [database (DB) schemas](https://en.wikipedia.org/wiki/Database_schema) to add columns noting when a resource was created or whether it was deleted. These columns can be used to automate data purges based on whether the data has been held longer than needed, for instance.

Reviewer: I'm not sure we need to link to Wikipedia here. Also, it seems like "Refactoring database schemas" might be a better heading.

Further DB schema analysis may also be required to ensure that more data isn't being collected than absolutely necessary.

## Compliance is worth the work

GDPR rules can be long and esoteric, but it's important for your organization to understand how to stay in compliance with the policy. Not only will you avoid fines for violations, but users will be more willing to trust your application. Ensuring better data privacy helps all stakeholders.

If necessary, remember you can seek help from a GDPR officer to ensure compliance. Doing this work upfront can save your organization from more costly consequences later on.

Check out these posts and whitepapers from Okta for more information about GDPR:

> - [Starting Your General Data Protection Regulation (GDPR) Journey with Okta](/whitepaper/starting-your-general-data-protection-regulation-journey-with-okta/)

Reviewer: This link is 404.

Author: all links seems to be 404, i'll correct them

- [GDPR, Privacy and Consent Management: How Okta Can Support Your CIAM Requirements](/blog/2019/01/gdpr-privacy-and-consent-management-how-okta-can-support-your-ciam-requirements/)
- [CCPA vs. GDPR: Similarities and Differences Explained](/blog/2021/04/ccpa-vs-gdpr/)

If you have any questions about this post, please add a comment below. For more interesting content, follow [@oktadev](https://twitter.com/oktadev) on Twitter, connect with us on [LinkedIn](https://www.linkedin.com/company/oktadev/), and subscribe to our [YouTube](https://www.youtube.com/oktadev) channel.

> Cover image credit: Photo of security cameras by Lianhao Qu via [Unsplash](https://unsplash.com/photos/LfaN1gswV5c)

Reviewer: ```suggestion
_Cover image credit: Photo of security cameras by Lianhao Qu via [Unsplash](https://unsplash.com/photos/LfaN1gswV5c)._
```

---

> type: awareness
> github: https://github.com/oktadev/okta-dotnet6-webapi-example

Reviewer: ```suggestion
type: awareness
```

## How GDPR affects application security

If you're responsible for managing your application's security, GDPR can feel like a threat. There is a lot of work involved, and mistakes can lead to costly fines. If you use it properly, though, following GDPR can help your application. Here are some of the benefits it can offer:

### Bigger security budget

Prior to GDPR, many organizations viewed cybersecurity as an expense rather than an investment. This attitude was reflected in understaffing and decreased purchasing power for critical cybersecurity features. The result was easily exploited digital infrastructure.

Because GDPR demands that organizations increase data security and backs up those demands with fines, managers of organizations would rather devote more of their budget to cybersecurity than risk losing profit to a security breach.

### More security integrations

> Applications that help with [automation testing](https://www.globalapptesting.com/blog/what-is-automation-testing), [automated breach and attack simulation](https://www.xmcyber.com/what-is-breach-and-attack-simulation/), [virtual private networks (VPNs)](https://www.kaspersky.com/resource-center/definitions/what-is-a-vpn), and [virtual private clouds (VPCs)](https://www.ibm.com/cloud/learn/vpc) are more likely to be added to an organization's security strategy. Such additions help shore up the defenses of digital infrastructure; VPCs and VPNs also offer a sandbox environment for employees to experiment with security features.

Reviewer 2: ```suggestion
Products and services that help with [automation testing](https://www.globalapptesting.com/blog/what-is-automation-testing), [automated breach and attack simulation](https://www.xmcyber.com/what-is-breach-and-attack-simulation/), [virtual private networks (VPNs)](https://www.kaspersky.com/resource-center/definitions/what-is-a-vpn), and [virtual private clouds (VPCs)](https://www.ibm.com/cloud/learn/vpc) are more likely to be added to an organization's security strategy. Such additions can shore up the defenses of digital infrastructure; VPCs and VPNs also offer a sandbox environment for employees to experiment with security features.
```
good call Matt

> ### Faster response to breaches

Reviewer 2: ```suggestion
### Faster response to security incidents
```
let's leave the word breach out of the heading

- Can users edit and update potentially inaccurate information?

**Data breach**

- If a data breach has occurred, have you informed supervisors within seventy-two hours of the breach?
- Have you carried out a data protection impact assessment before adopting new technology that interacts with user data?

**Data protection**

- Have you taken adequate measures to prevent the compromise of user data?
- Are your cybersecurity policies and measures up to date?
- Is personally identifiable information (PII) of users properly encrypted to prevent abuse by malignant actors?

**Data rights**

- Do users have the ability to request copies of their data?
- Do users have the ability to object to certain uses of their data?
- Can you delete a user's data immediately if they request it? Does this include removing their data from your backups?

**Data transfer**

- Is the data transfer process handled in a way that doesn't compromise data?
- Does the receiver have enough security installations in place to protect data?
- Does the receiver comply with the human rights charter?
- Is the receiver allowed to obtain data from entities bound by GDPR laws?

## Common GDPR features

> Following are examples of features that you can implement in your application to ensure GDPR compliance:

Reviewer 2: ```suggestion
Here are examples of features that you can implement in your application to ensure GDPR compliance:
```
avoid repetition

### Data mapping

[Data mapping](https://en.wikipedia.org/wiki/Data_mapping), or connecting information from multiple data sets, requires a top-down analysis of all of an organization's databases and what information they hold. This can be a big task, but it ensures transparency and accountability. Data mapping is required for the next steps.

Data mapping can be manual, semi-automated, or automated. You should determine which approach works best for you, depending on the size of your company, your budget, the quality of your data management team, and the amount of data you work with. Deciding how to [implement data mapping](https://blog.hubspot.com/marketing/data-mapping) will require some understanding of your databases, as well as knowing which duplicated columns can be optimized to reduce redundancy.

### Data purge functionality

Previously, even after users requested that their data be deleted, organizations could remove it from public access and keep it somewhere hidden; however, the GDPR mandates that "officially" deleted data must be deleted everywhere. Organizations must ensure that they monitor data to delete it properly from all resources as needed.

> Data purge functionality [implementation](https://success.outsystems.com/Documentation/Best_Practices/Architecture/Data_Purging) will likely require some modification to your database schema so that you can identify data to mark for deletion, per user requests, and data to purge due to staleness. A `LastUpdatedDate` and `IsDeleted` column in your data table can be useful in identifying either type of data.

Reviewer 2: ```suggestion
Implementing [data purge functionality](https://success.outsystems.com/Documentation/Best_Practices/Architecture/Data_Purging) will likely require some modification to your database schema so that you can identify data to mark for deletion, per user requests, and data to purge due to staleness.  Using `last_updated_date` and `is_deleted` columns in your data table can be useful in identifying either type of data.
```
applying Matt's suggestion

> ### Refactoring DB schemas

Reviewer 2: ```suggestion
### Refactoring database schemas
```
spells out database
